src/plugins/matrixops/concatenate.cpp

```diff
@@ -0,0 +1,817 @@
+            "phylanx::execution_tree::primitives::concatenate::concatenate3d_"
+            "helper",
+            generate_error_message("axis is out of bounds of dimension 3"));
+    }
+
+    primitive_argument_type concatenate::concatenate3d(
+        primitive_arguments_type&& args, std::int64_t axis) const
+    {
+        if (axis < -3 || axis > 2)
+            HPX_THROW_EXCEPTION(hpx::bad_parameter,
+                "concatenate::concatenate1d",
+                generate_error_message("axis is out of bounds of dimension 3"));
+
+        switch (extract_common_type(args))
+        {
+        case node_data_type_bool:
+            return concatenate3d_helper<std::uint8_t>(std::move(args), axis);
+
+        case node_data_type_int64:
+            return concatenate3d_helper<std::int64_t>(std::move(args), axis);
+
+        case node_data_type_unknown:
+            HPX_FALLTHROUGH;
+        case node_data_type_double:
+            return concatenate3d_helper<double>(std::move(args), axis);
+
+        default:
+            break;
+        }
+
+        HPX_THROW_EXCEPTION(hpx::bad_parameter,
+            "phylanx::execution_tree::primitives::concatenate::concatenate3d",
+            generate_error_message(
+                "the concatenate primitive requires for all arguments to "
+                "be numeric data types"));
+    }
+#endif
+    ///////////////////////////////////////////////////////////////////////////
+    hpx::future<primitive_argument_type> concatenate::eval(
+        primitive_arguments_type const& operands,
+        primitive_arguments_type const& args, eval_context ctx) const
+    {
+        if (operands.empty() || operands.size() > 2)
+        {
+            HPX_THROW_EXCEPTION(hpx::bad_parameter,
+                "concatenate::eval",
+                generate_error_message(
+                    "the concatenate primitive requires at most two operands"));
+        }
+
+        auto this_ = this->shared_from_this();
+        return hpx::dataflow(hpx::launch::sync,
+            hpx::util::unwrapping([this_ = std::move(this_)](
+                                      primitive_arguments_type&& args)
+                                      -> primitive_argument_type {
+                primitive_arguments_type ops;
+                if (!is_list_operand(args[0]))
+                    HPX_THROW_EXCEPTION(hpx::bad_parameter,
+                        "phylanx::execution_tree::primitives::"
+                        "concatenate::eval",
+                        this_->generate_error_message(
+                            "concatenate accepts a list "
+                            "value as its first operand"));
+                ir::range lhs = extract_list_value_strict(
+                    std::move(args[0]), this_->name_, this_->codename_);
+
+                if (lhs.is_ref())
+                {
+                    ops = lhs.copy();
+                }
+                else
+                    ops = lhs.args();
```

Comment:
Both conditional branches have the same effect, namely that the range is copied.